### DISCUSSION TITLE
A utility for converting composite OBJ files to a MuJoCo model

### DISCUSSION THREAD

**Discussion by kevinzakka** [google-deepmind/mujoco#340] [Show and tell]
Introducing [obj2mjcf](https://github.com/kevinzakka/obj2mjcf), a CLI for processing composite Wavefront [OBJ] files into a [MuJoCo]-conducive format. It automatically:

* Splits an OBJ file into sub-meshes that are grouped by the materials referenced in the OBJ's MTL file
* Creates a collision mesh by performing a convex decomposition of the OBJ with [V-HACD] (if available on your path)
* Generates an MJCF XML file that is pre-filled with materials, meshes and geom elements referencing these OBJ files

The generated meshes can then be used as a drop-in replacement for the original OBJ file. The result is vastly enhanced visuals for your model:

| Before | After |
|--------|-------|
|<img src="https://raw.githubusercontent.com/kevinzakka/obj2mjcf/main/assets/anymal_base_before.png" width="400"/>|<img src="https://raw.githubusercontent.com/kevinzakka/obj2mjcf/main/assets/anymal_base_after.png" width="400"/>|
|<img src="https://raw.githubusercontent.com/kevinzakka/obj2mjcf/main/assets/panda_link7_before.png" width="400"/>|<img src="https://raw.githubusercontent.com/kevinzakka/obj2mjcf/main/assets/panda_link7_after.png" width="400"/>|

**Why use this tool?** As of June 2022, MuJoCo does not support composite OBJ files consisting of groups or objects (`o` or `g` OBJ tags) and only 1 material can be assigned per mesh. This means that you have to manually split your OBJ file into sub-meshes, a process that is tedious and error-prone. This tool is meant to automate this process.

Happy simulating,
Kevin

[OBJ]: https://en.wikipedia.org/wiki/Wavefront_.obj_file
[MuJoCo]: https://github.com/deepmind/mujoco
[V-HACD]: https://github.com/kmammou/v-hacd

**Comment by yuvaltassa** [comment]
Thanks @kevinzakka, this is excellent!

In case someone asks the very legitimate question: "Will MuJoCo ever natively support loading **composite** OBJ files?" (as in discussion #255), the answer is no.

The reason for this is the following. MuJoCo loads meshes and textures separately, they have their own `<asset>` fields  and are considered to be different types of assets. It is only in the `<geom>` that the mesh and texture are referenced and combined, but a geom cannot itself be loaded as an asset as it is a model element.

So given the current relationship of geoms and assets, Kevin's excellent solution is also the optimal one.
